src/zw.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+//
+// This file contains Usersim overrides of some Zw* APIs
+// exposed by ntdll.dll, where the behavior of such APIs
+// differs between kernel mode and unprivileged user
+// mode, so that we can emulate kernel mode behavior to
+// unprivileged user mode test processes.
+#include "fault_injection.h"
+#include "usersim/zw.h"
+#include "utilities.h"
+#include <string>
+
+_IRQL_requires_max_(PASSIVE_LEVEL) NTSTATUS ZwCreateKey(
+    _Out_ PHANDLE key_handle,
+    _In_ ACCESS_MASK desired_access,
+    _In_ POBJECT_ATTRIBUTES object_attributes,
+    _Reserved_ ULONG title_index,
+    _In_opt_ PUNICODE_STRING class_string,
+    _In_ ULONG create_options,
+    _Out_opt_ PULONG disposition)
+{
+    if (usersim_fault_injection_inject_fault()) {
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    if (class_string != nullptr || title_index != 0) {
+        return STATUS_NOT_SUPPORTED;
+    }
+
+    UNREFERENCED_PARAMETER(create_options);
+
+    HKEY root_key = (HKEY)object_attributes->RootDirectory;
+    std::wstring relative_path = object_attributes->ObjectName->Buffer;
```

Comment:
Buffer is not guaranteed to be NULL terminated. Should use buffer + length to initialize string.